end_to_end_tests/generated_code_live_tests/test_unions.py

```diff
@@ -0,0 +1,233 @@
+      type: object
+      properties:
+        thing: {"$ref": "#/components/schemas/ThingAOrB"}
+      required: ["thing"]
+    ModelWithNestedUnion:
+      type: object
+      properties:
+        thingOrValue:
+          oneOf:
+            - "$ref": "#/components/schemas/ThingAOrB"
+            - oneOf:
+              - type: string
+              - type: number
+    ModelWithUnionOfOne:
+      type: object
+      properties:
+        thing:
+          oneOf:
+            - $ref: "#/components/schemas/ThingA"
+        requiredThing:
+          oneOf:
+            - $ref: "#/components/schemas/ThingA"
+      required: ["requiredThing"]
+""")
+@with_generated_code_imports(
+    ".models.ThingA",
+    ".models.ThingB",
+    ".models.ModelWithUnion",
+    ".models.ModelWithRequiredUnion",
+    ".models.ModelWithNestedUnion",
+    ".models.ModelWithUnionOfOne",
+    ".types.Unset"
+)
+class TestOneOf:
+    def test_disambiguate_objects_via_required_properties(self, ThingA, ThingB, ModelWithUnion):
+        assert_model_decode_encode(
+            ModelWithUnion,
+            {"thing": {"propA": "x"}},
+            ModelWithUnion(thing=ThingA(prop_a="x")),
+        )
+        assert_model_decode_encode(
+            ModelWithUnion,
+            {"thing": {"propB": "x"}},
+            ModelWithUnion(thing=ThingB(prop_b="x")),
+        )
+
+    def test_disambiguate_object_and_non_object(self, ThingA, ModelWithUnion):
+        assert_model_decode_encode(
+            ModelWithUnion,
+            {"thingOrString": {"propA": "x"}},
+            ModelWithUnion(thing_or_string=ThingA(prop_a="x")),
+        )
+        assert_model_decode_encode(
+            ModelWithUnion,
+            {"thingOrString": "x"},
+            ModelWithUnion(thing_or_string="x"),
+        )
+    
+    def test_disambiguate_nested_union(self, ThingA, ThingB, ModelWithNestedUnion):
+        assert_model_decode_encode(
+            ModelWithNestedUnion,
+            {"thingOrValue": {"propA": "x"}},
+            ModelWithNestedUnion(thing_or_value=ThingA(prop_a="x")),
+        )
+        assert_model_decode_encode(
+            ModelWithNestedUnion,
+            {"thingOrValue": 3},
+            ModelWithNestedUnion(thing_or_value=3),
+        )
+
+    def test_type_hints(self, ModelWithUnion, ModelWithRequiredUnion, ModelWithUnionOfOne, ThingA, Unset):
+        assert_model_property_type_hint(
+            ModelWithUnion,
+            "thing",
+            Union[ForwardRef("ThingA"), ForwardRef("ThingB"), Unset],
+        )
+        assert_model_property_type_hint(
+            ModelWithRequiredUnion,
+            "thing",
+            Union[ForwardRef("ThingA"), ForwardRef("ThingB")],
+        )
+        assert_model_property_type_hint(
+            ModelWithUnionOfOne, "thing", Union[ForwardRef("ThingA"), Unset]
+        )
+        assert_model_property_type_hint(
+            ModelWithUnionOfOne, "required_thing", "ThingA"
+        )
+
+
+@with_generated_client_fixture(
```

Comment:
The following test class covers all of the valid cases for discriminators. The error cases are in `test_invalid_unions.py`.